File: entities/bonfire.py
import pygame
import math

class Bonfire:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.width = 48  # Size of the bonfire
        self.height = 48
        
        # Create collision rect (slightly smaller than visual size for better gameplay)
        self.rect = pygame.Rect(self.x + 10, self.y + 20, self.width - 20, self.height - 20)
        
        # Animation variables
        self.frame_index = 0
        self.animation_speed = 0.15
        self.animation_timer = 0
        
        # Load sprites
        self.load_sprites()
    
    def load_sprites(self):
        """Load bonfire-specific sprites from sprite sheet"""
        self.sprites = {
            'burning': []  # Only one animation state for the bonfire
        }
        
        try:
            # Load the spritesheet
            bonfire_img = pygame.image.load('assets/bonfire-sprites.png').convert_alpha()
            
            # Define the frame positions based on the sprite sheet
            # These are the x-coordinates of each frame
            frame_positions = [
                0,    # Frame 1 start x-position
                32,   # Frame 2 start x-position
                64,   # Frame 3 start x-position
                96    # Frame 4 start x-position
            ]
            
            frame_width = 32   # Width of each frame
            frame_height = 32  # Height of each frame
            
            # Load each frame using defined positions
            for x_pos in frame_positions:
                frame = pygame.Surface((frame_width, frame_height), pygame.SRCALPHA)
                frame.blit(bonfire_img, (0, 0), (x_pos, 0, frame_width, frame_height))
                # Scale to desired size
                scaled_frame = pygame.transform.scale(frame, (self.width, self.height))
                self.sprites['burning'].append(scaled_frame)
                
        except Exception as e:
            print(f"Error loading bonfire sprites: {e}")
            # Create placeholder
            placeholder = pygame.Surface((self.width, self.height), pygame.SRCALPHA)
            placeholder.fill((200, 100, 0))
            self.sprites['burning'].append(placeholder)
    
    def get_rect(self):
        """Return the collision rectangle"""
        return self.rect
    
    def update(self, dt=None):
        """Update the bonfire animation"""
        # Track animation time
        self.animation_timer += 1
        
        # Update animation frame when timer exceeds threshold
        if self.animation_timer >= 10:  # Adjust this value to control animation speed
            self.frame_index = (self.frame_index + 1) % len(self.sprites['burning'])
            self.animation_timer = 0
    
    def get_current_frame(self):
        """Get the current animation frame"""
        return self.sprites['burning'][self.frame_index]
    
    def draw(self, surface):
        """Draw the bonfire with current animation frame"""
        # Get the current animation frame
        current_frame = self.get_current_frame()
        
        # Draw the bonfire
        surface.blit(current_frame, (self.x, self.y))

File: src/entities/bonfire.js
/**
 * Bonfire - Healing station and save point
 * Mirrors Python entities/bonfire.py
 */
class Bonfire {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.width = 48;
        this.height = 48;

        // Bonfire is an obstacle - player cannot walk through it
        this.isObstacle = true;

        // Animation
        this.frameIndex = 0;
        this.animationTimer = 0;
        this.frameCount = 4;
        this.animationSpeed = 10; // Frames between animation updates

        // Healing effect
        this.healParticlesActive = false;
        this.healParticlesTimer = 0;

        // Block coordinates
        this.blockX = 0;
        this.blockY = 0;

        // Save/load callback (set by main.js for origin bonfire)
        this.saveLoadCallback = null;

        // Load sprites
        this.sprites = [];
        this.loadSprites();
    }

    async loadSprites() {
        try {
            const img = new Image();
            img.src = 'assets/bonfire-sprites.png';

            await new Promise((resolve, reject) => {
                img.onload = resolve;
                img.onerror = reject;
            });

            // Frame positions in the sprite sheet
            const frameWidth = 32;
            const frameHeight = 32;
            const framePositions = [0, 32, 64, 96];

            // Create offscreen canvas for each frame
            for (const xPos of framePositions) {
                const canvas = document.createElement('canvas');
                canvas.width = this.width;
                canvas.height = this.height;
                const ctx = canvas.getContext('2d');

                // Draw and scale frame
                ctx.drawImage(
                    img,
                    xPos, 0, frameWidth, frameHeight,
                    0, 0, this.width, this.height
                );

                this.sprites.push(canvas);
            }

            console.log('Bonfire sprites loaded');
        } catch (e) {
            console.log('Could not load bonfire sprites, using placeholder');
            // Create placeholder
            const canvas = document.createElement('canvas');
            canvas.width = this.width;
            canvas.height = this.height;
            const ctx = canvas.getContext('2d');
            ctx.fillStyle = 'rgb(200, 100, 0)';
            ctx.fillRect(0, 0, this.width, this.height);
            this.sprites = [canvas];
        }
    }

    getRect() {
        return {
            x: this.x,
            y: this.y,
            width: this.width,
            height: this.height
        };
    }

    setBlockCoordinates(x, y) {
        this.blockX = x;
        this.blockY = y;
    }

    isOriginBonfire() {
        return this.blockX === 0 && this.blockY === 0;
    }

    update(dt, currentTime) {
        // Update animation
        this.animationTimer++;
        if (this.animationTimer >= this.animationSpeed) {
            this.frameIndex = (this.frameIndex + 1) % (this.sprites.length || 1);
            this.animationTimer = 0;
        }

        // Update heal particles
        if (this.healParticlesActive) {
            this.healParticlesTimer++;
            if (this.healParticlesTimer > 60) {
                this.healParticlesActive = false;
                this.healParticlesTimer = 0;
            }
        }
    }

    interact(player, currentTime) {
        console.log('Bonfire interact method called!');

        // Calculate heal amount
        const healAmount = player.attributes.maxHealth - player.attributes.currentHealth;

        console.log(`Player health before: ${player.attributes.currentHealth}/${player.attributes.maxHealth}`);

        let healed = false;
        if (healAmount > 0) {
            player.heal(healAmount);
            this.healParticlesActive = true;
            this.healParticlesTimer = 0;
            console.log(`Bonfire healed player for ${healAmount} HP`);
            healed = true;
        } else {
            console.log('Player already at full health');
        }

        // Restore mana to full as well
        const manaAmount = player.attributes.maxMana - player.attributes.currentMana;
        if (manaAmount > 0) {
            player.attributes.restoreMana(manaAmount);
            console.log(`Bonfire restored ${manaAmount} mana`);
        }

        // If this is the origin bonfire, show save/load dialog
        if (this.isOriginBonfire() && this.saveLoadCallback) {
            console.log('Origin bonfire - showing save/load dialog');
            this.saveLoadCallback();
        }

        return healed || this.isOriginBonfire();
    }

    render(ctx, game) {
        // Draw current frame
        if (this.sprites.length > 0) {
            const frame = this.sprites[this.frameIndex % this.sprites.length];
            ctx.drawImage(frame, this.x, this.y);
        } else {
            // Placeholder
            ctx.fillStyle = 'rgb(255, 100, 0)';
            ctx.fillRect(this.x, this.y, this.width, this.height);
        }

        // Draw heal particles if active
        if (this.healParticlesActive) {
            const centerX = this.x + this.width / 2;
            const centerY = this.y + this.height / 2;

            for (let i = 0; i < 8; i++) {
                const particleSize = 3 + (this.healParticlesTimer % 3);
                const angle = i * Math.PI / 4;
                const distance = 20 + this.healParticlesTimer / 5;
                const particleX = centerX + Math.cos(angle) * distance;
                const particleY = centerY + Math.sin(angle) * distance - this.healParticlesTimer / 3;

                ctx.fillStyle = 'rgb(100, 255, 100)';
                ctx.beginPath();
                ctx.arc(particleX, particleY, particleSize, 0, Math.PI * 2);
                ctx.fill();
            }
        }

        // Draw debug collision box when C is held
        if (game && game.showDebug) {
            const rect = this.getRect();
            ctx.strokeStyle = '#00ff00';
            ctx.lineWidth = 2;
            ctx.strokeRect(rect.x, rect.y, rect.width, rect.height);
        }
    }
}

// Export
window.Bonfire = Bonfire;
